src/examples/libpmem2/unsafe_shutdowns/unsafe_shutdowns.c

```diff
@@ -0,0 +1,607 @@
+
+struct pool_state {
+	struct device_state {
+		char id[DEVICE_ID_LEN];
+		uint64_t usc; /* unsafe shutdown counter value */
+	} device_state;
+
+	uint8_t file_in_use;
+};
+
+struct pool_content {
+	struct pool_state ps;
+	char usable_data[];
+};
+
+struct pool_data {
+	struct pool_content *content;
+
+	/* source and mapping objects */
+	struct pmem2_source *src;
+	struct pmem2_map *map;
+
+	size_t map_size;
+	size_t usable_space_size;
+
+	/* mapping-specific functions */
+	pmem2_persist_fn persist;
+	pmem2_memset_fn memset;
+	pmem2_memcpy_fn memcpy;
+};
+
+/*
+ * device_state_read -- read device UUID and USC
+ */
+static int
+device_state_read(struct pmem2_source *src, struct device_state *ds)
+{
+	/* obtain device unsafe shutdown counter value (USC) */
+	int ret = pmem2_source_device_usc(src, &ds->usc);
+	if (ret) {
+		pmem2_perror("pmem2_source_device_usc");
+		return ret;
+	}
+
+	/* obtain a device's ID */
+	size_t len = 0;
+	ret = pmem2_source_device_id(src, NULL, &len);
+	if (ret) {
+		pmem2_perror(
+			"pmem2_source_device_id failed querying device ID length");
+		return ret;
+	}
+
+	if (len > DEVICE_ID_LEN) {
+		fprintf(stderr, "the device ID is too long (%zu > %zu)\n",
+			len, DEVICE_ID_LEN);
+		return 1;
+	}
+
+	ret = pmem2_source_device_id(src, (char *)&ds->id, &len);
+	if (ret) {
+		pmem2_perror("pmem2_source_device_id failed reading device ID");
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * device_state_init -- initialize device state in a unsafe-shutdown-safe way
+ */
+static int
+device_state_init(struct pool_data *pool)
+{
+	struct pool_state *ps = &pool->content->ps;
+	struct device_state *ds = &ps->device_state;
+
+	/* validate initial pool state */
+	const struct pool_state ps_zeroed = {0};
+	if (memcmp(&pool->content->ps, &ps_zeroed, sizeof(ps_zeroed)) != 0) {
+		fprintf(stderr, "the file is not zero-initialized\n");
+		return 1;
+	}
+
+	/* read current device state */
+	struct device_state ds_curr = {0};
+	int ret = device_state_read(pool->src, &ds_curr);
+
+	/*
+	 * write, persist and deep sync the USC value. It has to be stored on
+	 * the persistent medium before it will be validated by writing device
+	 * ID.
+	 */
+	ds->usc = ds_curr.usc;
+	pool->persist(&ds->usc, sizeof(ds->usc));
+	ret = pmem2_deep_sync(pool->map, &ds->usc, sizeof(ds->usc));
+	if (ret) {
+		pmem2_perror("pmem2_deep_sync USC deep sync failed");
+		return ret;
+	}
+
+	/* valid device ID validates already stored USC value */
+	pool->memcpy(ds->id, ds_curr.id, DEVICE_ID_LEN, FLAGS_ZERO);
+	ret = pmem2_deep_sync(pool->map, ds->id, DEVICE_ID_LEN);
+	if (ret) {
+		pmem2_perror("pmem2_deep_sync device ID failed");
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * device_state_reinit -- reinitialize device state
+ * This is required in a few cases:
+ * - the primary initialization was interrupted leaving the pool half-backed
+ * - the pool was moved which invalidates all collected device-specific data
+ */
+static int
+device_state_reinit(struct pool_data *pool)
```

Comment:
device_state_reset()

This should be then explicitly followed by init() in the calling function.

And again, this function should take device_state.